### ISSUE TITLE
error pushing image: failed to push to destination with Amazon ECR

### ISSUE THREAD

**Issue by didiwu456** [triggermesh/knative-lambda-runtime#60]
@tzununbekov I tried to use KLR with Amazon EKS and Amazon ECR but received the following error message when trying to deploy the sample Python function. Could you let me know if there's any documentation about this?
> INFO[0000] Downloading base image gcr.io/triggermesh/knative-lambda-python37
> 2021/03/25 20:00:35 Unable to read "/config.json": open /config.json: no such file or directory
> 2021/03/25 20:00:35 No matching credentials were found, falling back on anonymous
> INFO[0000] Unpacking rootfs as cmd RUN if [ -f requirements.txt ]; then pip3.7 install -r requirements.txt ;fi requires it.
> INFO[0005] Taking snapshot of full filesystem...
> INFO[0006] Skipping paths under /kaniko, as it is a whitelisted directory
> INFO[0006] Skipping paths under /workspace, as it is a whitelisted directory
> INFO[0006] Skipping paths under /tekton/tools, as it is a whitelisted directory
> INFO[0006] Skipping paths under /tekton/home, as it is a whitelisted directory
> INFO[0006] Skipping paths under /tekton/results, as it is a whitelisted directory
> INFO[0006] Skipping paths under /tekton/creds, as it is a whitelisted directory
> INFO[0006] Skipping paths under /var/run, as it is a whitelisted   @@directory
> INFO[0006] Skipping paths under /dev, as it is a whitelisted directory
> INFO[0006] Skipping paths under /proc, as it is a whitelisted directory
> INFO[0006] Skipping paths under /sys, as it is a whitelisted directory
> INFO[0006] Skipping paths under /busybox, as it is a whitelisted directory
> INFO[0007] ENV _HANDLER "handler.endpoint"
> INFO[0007] Using files from context: [/workspace/workspace/aws-python-simple-http-endpoint]
> INFO[0007] COPY . .
> INFO[0007] Taking snapshot of files...
> INFO[0007] RUN if [ -f requirements.txt ]; then pip3.7 install -r requirements.txt ;fi
> INFO[0007] cmd: /bin/sh
> INFO[0007] args: [-c if [ -f requirements.txt ]; then pip3.7 install -r requirements.txt ;fi]
> INFO[0007] Taking snapshot of full filesystem...
> INFO[0007] Skipping paths under /kaniko, as it is a whitelisted directory
> INFO[0007] Skipping paths under /workspace, as it is a whitelisted directory
> INFO[0007] Skipping paths under /tekton/tools, as it is a whitelisted directory
> INFO[0007] Skipping paths under /tekton/home, as it is a whitelisted directory
> INFO[0007] Skipping paths under /tekton/results, as it is a whitelisted directory
> INFO[0007] Skipping paths under /tekton/creds, as it is a whitelisted directory
> INFO[0007] Skipping paths under /var/run, as it is a whitelisted directory
> INFO[0007] Skipping paths under /dev, as it is a whitelisted directory
> INFO[0007] Skipping paths under /proc, as it is a whitelisted directory
> INFO[0007] Skipping paths under /sys, as it is a whitelisted directory
> INFO[0007] Skipping paths under /busybox, as it is a whitelisted directory
> INFO[0008] No files were changed, appending empty layer to config. No layer added to image.
> INFO[0008] ENTRYPOINT ["/opt/aws-custom-runtime"]
> 2021/03/25 20:00:44 Unable to read "/config.json": open /config.json: no such file or directory
> error pushing image: failed to push to destination index.docker.io/ecr/default/python-test:LRM892: unsupported status code 401; body:

**Comment by tzununbekov** [comment]
@didiwu456 from the log you've posted I see that the resulting image is being pushed to the `index.docker.io/ecr/default/python-test` destination, which doesn't seem right to me. There is a readme section in tm CLI repository that describes the registry usage [here](https://github.com/triggermesh/tm#docker-registry). What commands did you use to configure the registry and deploy KLR?

**Comment by didiwu456** [comment]
@tzununbekov I first get the login password from AWS CLI.
`aws ecr get-login-password --region my-region`
Use the password to set registry.
```
tm set registry-auth ecr
Registry: my-account.dkr.ecr.my-region.amazonaws.com
Username: AWS
Password:
```
Then, I ran this command.
`tm deploy service python-test -f https://github.com/serverless/examples --build-template knative-python37-runtime --build-argument DIRECTORY=aws-python-simple-http-endpoint --build-argument HANDLER=handler.endpoint --wait --registry-host ecr`

**Comment by tzununbekov** [comment]
@didiwu456 commands you use look good, but I suspect that your CLI binary is outdated. Please make sure that you use the [latest](https://github.com/triggermesh/tm/releases/latest) version of CLI. Also, after updating the CLI could you run `tm deploy service ...` command with an additional `-d` argument to see the debug output (and post it here if it still doesn't work)?

**Comment by didiwu456** [comment]
@tzununbekov I've executed `tm deploy service python-test -f https://github.com/serverless/examples --build-template knative-pt --wait --registry-host ecr -d`. Please check the output below:
```
time="2021-03-29T09:00:57-07:00" level=debug msg="taskrun \"default/python-test\" output image will be \"ecr/default/python-test:77iI8E\""
time="2021-03-29T09:00:57-07:00" level=debug msg="taskrun \"default/python-test-78bpr\" created"
time="2021-03-29T09:00:57-07:00" level=debug msg="setting pipelineresource owner"
time="2021-03-29T09:00:57-07:00" level=info msg="Waiting for taskrun \"python-test-78bpr\" ready state"
time="2021-03-29T09:00:57-07:00" level=debug msg="got new event:"
time="2021-03-29T09:00:58-07:00" level=debug msg="got new event:"
time="2021-03-29T09:00:58-07:00" level=debug msg="got new event:"
time="2021-03-29T09:00:58-07:00" level=debug msg=" condition: \"Succeeded\", status: \"Unknown\", message: \"Pending\""
time="2021-03-29T09:00:58-07:00" level=debug msg="got new event:"
time="2021-03-29T09:00:58-07:00" level=debug msg=" condition: \"Succeeded\", status: \"Unknown\", message: \"pod status \\\"Initialized\\\":\\\"False\\\"; message: \\\"containers with incomplete status: [place-tools]\\\"\""
time="2021-03-29T09:01:00-07:00" level=debug msg="got new event:"
time="2021-03-29T09:01:00-07:00" level=debug msg=" condition: \"Succeeded\", status: \"Unknown\", message: \"pod status \\\"Ready\\\":\\\"False\\\"; message: \\\"containers with unready status: [step-git-source-sources-bbpwc step-dockerfile step-export]\\\"\""
time="2021-03-29T09:01:01-07:00" level=debug msg="got new event:"
time="2021-03-29T09:01:01-07:00" level=debug msg=" condition: \"Succeeded\", status: \"Unknown\", message: \"Not all Steps in the Task have finished executing\""
time="2021-03-29T09:01:05-07:00" level=debug msg="got new event:"
time="2021-03-29T09:01:05-07:00" level=debug msg=" condition: \"Succeeded\", status: \"Unknown\", message: \"Not all Steps in the Task have finished executing\""
time="2021-03-29T09:01:15-07:00" level=debug msg="got new event:"
time="2021-03-29T09:01:15-07:00" level=debug msg=" condition: \"Succeeded\", status: \"False\", message: \"\\\"step-export\\\" exited with code 1 (image: \\\"docker-pullable://gcr.io/kaniko-project/executor@sha256:a49506bad46fa4a8ce412239b25ebc2577290cc21ca04a0618d6157e379f811e\\\"); for logs run: kubectl -n default logs python-test-78bpr-pod-552kd -c step-export\\n\""
time="2021-03-29T09:01:15-07:00" level=debug msg="setting taskrun \"default/python-test-78bpr\" owner to Configuration/"
time="2021-03-29T09:01:15-07:00" level=fatal msg="Deploying builder: taskrun \"python-test-78bpr\" deployment failed: \"step-export\" exited with code 1 (image: \"docker-pullable://gcr.io/kaniko-project/executor@sha256:a49506bad46fa4a8ce412239b25ebc2577290cc21ca04a0618d6157e379f811e\"); for logs run: kubectl -n default logs python-test-78bpr-pod-552kd -c step-export\n"
```
Also, The version of CLI is `tm version v1.3.0`

**Comment by tzununbekov** [comment]
@didiwu456 I think I figured out what is wrong here. The first issue is that in `tm deploy service ...` command we use `--registry-host ecr` parameter which actually sets the **registry host** to "ecr". In order to make deployment use the **registry secret** we should pass ` --registry-secret ecr` parameter. The second thing that I bumped into after fixing the registry parameter is that ECR requires to have all repositories created before pushing images, i.e. you cannot `docker push ecr.amazonaws.com/bar/foo:latest` and expect that this repository will be created automatically. Here is how KLR deployment can be done with ECR:
1. Create ECR repository in `<project>/<service>` format. In our case, it should be  `triggermesh/python-test`
2. Get the ECR token: `aws ecr get-login-password ...`
3. `tm set registry-auth ecr --registry ***.amazonaws.com --project "triggermesh" --username "AWS"`. Insert the ECR token.
4. Finally deploy the service: 
```
tm deploy service python-test -f https://github.com/serverless/examples \
  --runtime https://raw.githubusercontent.com/triggermesh/knative-lambda-runtime/master/python37/runtime.yaml \
  --registry-secret ecr \
  --build-argument DIRECTORY=aws-python-simple-http-endpoint \
  --build-argument HANDLER=handler.endpoint \
  --wait
```
Let me know if it worked for you
